[JSC] Generic iterator protocol function should be invisible to Function's caller getter

Constellation opened the pull request:
#### 1a1af3bd846d98207ea02c8aa079351347a55205
<pre>
[JSC] Generic iterator protocol function should be invisible to Function&apos;s caller getter
<a href="https://bugs.webkit.org/show_bug.cgi?id=241953">https://bugs.webkit.org/show_bug.cgi?id=241953</a>

Reviewed by Alexey Shvayka.

When we use iterator protocol helper function, `caller` getter inside it can encounter this
function, and returning null because it is builtin function. Now, we leverage ImplementationVisibility
to skip iteration protocol helper function from `caller` getter&apos;s stack walk so that we can make it
invisible from user&apos;s code.

* JSTests/stress/iteration-helper-hidden.js: Added.
(shouldBe):
(foo):
(obj.Symbol.iterator):
* Source/JavaScriptCore/builtins/IteratorHelpers.js:
(performIteration): Deleted.
* Source/JavaScriptCore/runtime/FunctionPrototype.cpp:
(JSC::isAllowedReceiverFunctionForCallerAndArguments):
(JSC::RetrieveCallerFunctionFunctor::operator() const):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::iteratorProtocolFunction const): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::iteratorProtocolFunction const):

Canonical link: <a href="https://commits.webkit.org/252578@main">https://commits.webkit.org/252578@main</a>
</pre>

shvaikalesh commented on Source/JavaScriptCore/runtime/FunctionPrototype.cpp:
```diff
@@ -108,6 +109,8 @@ static ALWAYS_INLINE bool isAllowedReceiverFunctionForCallerAndArguments(JSFunct
         return false;
 
     FunctionExecutable* executable = function->jsExecutable();
+    if (executable->implementationVisibility() == ImplementationVisibility::Private)
```
Can this check even pass after

```
if (function->isHostOrBuiltinFunction())
         return false;
```

?

Maybe we should remove it, as builtin functions can't be receivers for `caller`?

Constellation replied:
Yes, non-builtin function can be Private (e.g. class static initializer function).

shvaikalesh commented on Source/JavaScriptCore/runtime/FunctionPrototype.cpp:
```diff
@@ -193,8 +193,19 @@ class RetrieveCallerFunctionFunctor {
+                auto* function = jsCast<JSFunction*>(callee);
+                if (function->executable()->implementationVisibility() == ImplementationVisibility::Private)
+                    return IterationStatus::Continue;
+                if (function->isHostOrBuiltinFunction())
```
Do we actually need this check? If we return a `function` here that is `isHostOrBuiltinFunction`, it will be rejected a bit later in `callerGetter`:

```
    if (!function || function->isHostOrBuiltinFunction())
        return JSValue::encode(jsNull());
```

I find it nicer to separate call frame iteration where we `IterationStatus::Continue` or not with vetting the final result, so maybe it's worth to have `isHostOrBuiltinFunction` only in `callerGetter`?

Constellation replied:
That's nice. Dropped.

Constellation committed:
[JSC] Generic iterator protocol function should be invisible to Function's caller getter
https://bugs.webkit.org/show_bug.cgi?id=241953

Reviewed by Alexey Shvayka.

When we use iterator protocol helper function, `caller` getter inside it can encounter this
function, and returning null because it is builtin function. Now, we leverage ImplementationVisibility
to skip iteration protocol helper function from `caller` getter's stack walk so that we can make it
invisible from user's code.

* JSTests/stress/iteration-helper-hidden.js: Added.
(shouldBe):
(foo):
(obj.Symbol.iterator):
* Source/JavaScriptCore/builtins/IteratorHelpers.js:
(performIteration): Deleted.
* Source/JavaScriptCore/runtime/FunctionPrototype.cpp:
(JSC::isAllowedReceiverFunctionForCallerAndArguments):
(JSC::RetrieveCallerFunctionFunctor::operator() const):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::iteratorProtocolFunction const): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::iteratorProtocolFunction const):

Canonical link: https://commits.webkit.org/252578@main
